src/java.base/share/classes/java/net/Socket.java

```diff
@@ -1095,6 +1098,20 @@ public int read() throws IOException {
         }
         @Override
         public int read(byte[] b, int off, int len) throws IOException {
+            if (! SocketReadEvent.enabled()) {
+                return read0(b, off, len);
+            }
+            int nbytes = 0;
+            long start = SocketReadEvent.timestamp();
+            try {
+                nbytes = read0(b, off, len);
+            } finally {
+                SocketReadEvent.checkForCommit(start, nbytes, parent.getRemoteSocketAddress(), parent.getSoTimeout());
```

Comment:
The other issue to think about here is where the Socket is asynchronously closed. In t hat case, implRead will throw but we'll end up with a confusing suppressed exception due to the call to getSoTimeout. I think this will have to be replaced with a call to a helper method that returns the timeout or 0.